microbench-ci: compare only outer run when insignificant

herkolategan opened the pull request:
Previously, during the comparison step all runs were compared. Since the last run determines if the previous runs had significant changes, we now first compare only the last outer run of each benchmark. If the last run had significant changes, we then compare all runs to produce the final assessment.

This prevents having a compare summary on CI that shows a regressions, when in fact only one run possibly had a regression, followed by an insignificant change in the last run.

i.e.,

Did the last run (3rd out of 3 runs) contain a regression / improvement?

*yes*: then implicitly all previous runs contained a regression or improvement, and should be included in the result summary.

*no*: then it's possible a previous run did regress, but we're disregarding the result in any case (due to requiring consecutive regressions in totality); hence only present the last insignificant run in the summary, to avoid confusion.

Epic: None
Release note: None

herkolategan committed:
microbench-ci: compare only outer run when insignificant

Previously, during the comparison step all runs were compared. Since the last
run determines if the previous runs had significant changes, we now first
compare only the last outer run of each benchmark. If the last run had
significant changes, we then compare all runs to produce the final assessment.

This prevents having a compare summary on CI that shows a regressions, when in
fact only one run possibly had a regression, followed by an insignificant change
in the last run.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143061)
<!-- Reviewable:end -->

herkolategan committed:
microbench-ci: fix status reporting

Previously, in some cases we did not set `NoChange` if the benchmark was
unchanged. The regression marker file should also only be written to the New
revision.

herkolategan commented:
TFTR!

bors r=DarrylWong
